File: p5/src/world.ts
import p5 from "p5";
import {
  Position,
  Velocity,
  Circle,
  GravitationalPull,
  BounceBoundaries,
  LeadMovement,
  DistanceConstraint,
  ChainLink,
  Snake,
  Orientation,
} from "./components/index";
import { i, s } from "../node_modules/vite/dist/node/types.d-aGj9QkWt";

interface System {
  update: () => void;
}

interface RenderSystem {
  draw: (p: p5, op: any) => void;
}

// Move to specific types
type ComponentsList = {
  initialPosition?: Position;
  position?: Position;
  velocity?: Velocity;
  circle?: Circle;
  gravitationalPull?: GravitationalPull;
  bounceBoundaries?: BounceBoundaries;
  leadMovement?: LeadMovement;
  distanceConstraint?: DistanceConstraint;
  chainLink?: ChainLink;
  snake?: Snake;
  orientation?: Orientation;
};

export interface World {
  p: p5;
  setup: () => void;
  draw: () => void;
  update: () => void;
  addEntity: () => number;
  addComponents: (entity: number, components: ComponentsList) => void;
  reset: () => void;
  addSystem: (system: System) => void;
  addRenderSystem: (system: RenderSystem) => void;

  // Components
  positions: { [key: number]: Position };
  initialPositions: { [key: number]: Position };
  velocities: { [key: number]: Velocity };
  circles: { [key: number]: Circle };
  gravitationalPulls: { [key: number]: GravitationalPull };
  bounceBoundaries: { [key: number]: BounceBoundaries };
  leadMovements: { [key: number]: LeadMovement };
  distanceConstraints: { [key: number]: DistanceConstraint };
  chainLinks: { [key: number]: ChainLink };
  snakes: { [key: number]: Snake };
  orientations: { [key: number]: Orientation };
}

export const newWorld = (p: p5): World => {
  const systems: System[] = [];
  const renderSystems: RenderSystem[] = [];
  let entityId = 0;
  let positions: { [key: number]: Position } = {};
  let initialPositions: { [key: number]: Position } = {};
  let velocities: { [key: number]: Velocity } = {};
  let circles: { [key: number]: Circle } = {};
  let gravitationalPulls: { [key: number]: GravitationalPull } = {};
  let bounceBoundaries: { [key: number]: BounceBoundaries } = {};
  let leadMovements: { [key: number]: LeadMovement } = {};
  let distanceConstraints: { [key: number]: DistanceConstraint } = {};
  let chainLinks: { [key: number]: ChainLink } = {};
  let snakes: { [key: number]: Snake } = {};
  let orientations: { [key: number]: Orientation } = {};

  const update = () => {
    for (const system of systems) {
      system.update();
    }
  };

  return {
    p,
    positions,
    initialPositions,
    velocities,
    circles,
    gravitationalPulls,
    bounceBoundaries,
    leadMovements,
    distanceConstraints,
    chainLinks,
    snakes,
    orientations,
    setup: () => {
      p.setup = () => {
        p.createCanvas(1600, 1200);
      };
    },
    reset: () => {
      entityId = 0;
      positions = {};
      initialPositions = {};
      velocities = {};
      circles = {};
      gravitationalPulls = {};
      bounceBoundaries = {};
      leadMovements = {};
      distanceConstraints = {};
      chainLinks = {};
      snakes = {};
      orientations = {};
    },
    draw: () => {
      p.draw = () => {
        update();
        p.background(0);
        for (const system of renderSystems) {
          system.draw(p, {});
        }
      };
    },
    update,
    addEntity: () => {
      return entityId++;
    },

    addSystem: (system: System) => {
      systems.push(system);
    },
    addRenderSystem: (system: RenderSystem) => {
      renderSystems.push(system);
    },
    addComponents: (entity: number, components: ComponentsList) => {
      if (components.position) {
        positions[entity] = components.position;
      }
      if (components.velocity) {
        velocities[entity] = components.velocity;
      }
      if (components.circle) {
        circles[entity] = components.circle;
      }
      if (components.gravitationalPull) {
        gravitationalPulls[entity] = components.gravitationalPull;
      }
      if (components.bounceBoundaries) {
        bounceBoundaries[entity] = components.bounceBoundaries;
      }
      if (components.leadMovement) {
        leadMovements[entity] = components.leadMovement;
      }
      if (components.distanceConstraint) {
        distanceConstraints[entity] = components.distanceConstraint;
      }
      if (components.chainLink) {
        chainLinks[entity] = components.chainLink;
      }
      if (components.snake) {
        snakes[entity] = components.snake;
      }
      if (components.orientation) {
        orientations[entity] = components.orientation;
      }
      if (components.initialPosition) {
        initialPositions[entity] = components.initialPosition;
      }
    },
  };
};

File: ebiten/src/pkg/world/world.go
package world

import (
	"github.com/hajimehoshi/ebiten/v2"
	"github.com/tonitienda/procedural-animations-go/src/pkg/components"
	"github.com/tonitienda/procedural-animations-go/src/pkg/entities"
)

type Poins struct {
	X float64
	Y float64
}

// FIXME - entities map should be private or readonly
type World struct {
	entities            []entities.Entity
	Positions           map[entities.Entity]*components.Position
	InitialPositions    map[entities.Entity]*components.InitialPosition
	Velocities          map[entities.Entity]*components.Velocity
	GravitationalPulls  map[entities.Entity]*components.GravitationalPull
	Circles             map[entities.Entity]*components.Circle
	BoundaryBouncings   map[entities.Entity]*components.BounceBoundaries
	LeadMovements       map[entities.Entity]*components.LeadMovement
	DistanceConstraints map[entities.Entity]*components.DistanceConstraint
	Snakes              map[entities.Entity]*components.Snake
	Orientations        map[entities.Entity]*components.Orientation
	systems             []System
	renderSystems       []RenderSystem
	nextEntityID        entities.Entity
}

type System interface {
	Update()
}

type DrawOptions struct {
	Debug bool
}

type RenderSystem interface {
	Draw(screen *ebiten.Image, op *DrawOptions)
}

func NewWorld() *World {
	return &World{
		Positions:           make(map[entities.Entity]*components.Position),
		InitialPositions:    make(map[entities.Entity]*components.InitialPosition),
		Velocities:          make(map[entities.Entity]*components.Velocity),
		GravitationalPulls:  make(map[entities.Entity]*components.GravitationalPull),
		Circles:             make(map[entities.Entity]*components.Circle),
		BoundaryBouncings:   make(map[entities.Entity]*components.BounceBoundaries),
		LeadMovements:       make(map[entities.Entity]*components.LeadMovement),
		DistanceConstraints: make(map[entities.Entity]*components.DistanceConstraint),
		Orientations:        make(map[entities.Entity]*components.Orientation),
		Snakes:              make(map[entities.Entity]*components.Snake),
		systems:             []System{},
		nextEntityID:        0,
	}
}

func (w *World) AddEntity() entities.Entity {
	id := w.nextEntityID
	w.nextEntityID++
	w.entities = append(w.entities, id)
	return id
}

func (w *World) AddComponents(entity entities.Entity, args ...interface{}) {
	for _, component := range args {
		switch c := component.(type) {
		case *components.Position:
			w.Positions[entity] = c
		case *components.InitialPosition:
			w.InitialPositions[entity] = c
		case *components.GravitationalPull:
			w.GravitationalPulls[entity] = c
		case *components.Circle:
			w.Circles[entity] = c
		case *components.Velocity:
			w.Velocities[entity] = c
		case *components.BounceBoundaries:
			w.BoundaryBouncings[entity] = c
		case *components.LeadMovement:
			w.LeadMovements[entity] = c
		case *components.DistanceConstraint:
			w.DistanceConstraints[entity] = c
		case *components.Snake:
			w.Snakes[entity] = c
		case *components.Orientation:
			w.Orientations[entity] = c
		default:
			// Handle unknown component types if necessary
		}
	}
}

func (w *World) AddSystem(system System) {
	w.systems = append(w.systems, system)
}

func (w *World) AddRenderSystem(system RenderSystem) {
	w.renderSystems = append(w.renderSystems, system)
}

func (w *World) Update() {
	for _, system := range w.systems {
		system.Update()
	}
}

// FIXME - Workd should be independent of the rendering system
func (w *World) Draw(screen *ebiten.Image, op *DrawOptions) {
	for _, system := range w.renderSystems {
		system.Draw(screen, op)
	}
}

func (w *World) Reset() {
	w.systems = []System{}
	w.Positions = make(map[entities.Entity]*components.Position)
	w.Velocities = make(map[entities.Entity]*components.Velocity)
	w.GravitationalPulls = make(map[entities.Entity]*components.GravitationalPull)
	w.Circles = make(map[entities.Entity]*components.Circle)
	w.BoundaryBouncings = make(map[entities.Entity]*components.BounceBoundaries)
	w.LeadMovements = make(map[entities.Entity]*components.LeadMovement)
	w.DistanceConstraints = make(map[entities.Entity]*components.DistanceConstraint)
	w.Snakes = make(map[entities.Entity]*components.Snake)
	w.nextEntityID = 0
}
